src/buildHTML.js

```diff
@@ -88,146 +64,173 @@ export const buildExpression = function(
+    // to avoid processing groups multiple times.
+    if (!isRealGroup) {
+        return groups;
+    }
+
+    let glueOptions = options;
+    if (expression.length === 1) {
+        const node = checkNodeType(expression[0], "sizing") ||
+            checkNodeType(expression[0], "styling");
+        if (!node) {
+            // No match.
+        } else if (node.type === "sizing") {
+            glueOptions = options.havingSize(node.size);
+        } else if (node.type === "styling") {
+            glueOptions = options.havingStyle(styleMap[node.style]);
         }
     }
-    // At this point `rawGroups` consists entirely of `symbolNode`s and `span`s.
 
-    // Ignore explicit spaces (e.g., \;, \,) when determining what implicit
-    // spacing should go between atoms of different classes, and add dummy
-    // spans for determining spacings between surrounding atoms.
-    const nonSpaces: (?HtmlDomNode)[] = [
-        surrounding[0] ? makeSpan([surrounding[0]], [], options) : null,
-        ...rawGroups.filter(group => group && group.classes[0] !== "mspace"),
-        surrounding[1] ? makeSpan([surrounding[1]], [], options) : null,
-    ];
+    // Dummy spans for determining spacings between surrounding atoms.
+    // If `expression` has no atoms on the left or right, class "leftmost"
+    // or "rightmost", respectively, is used to indicate it.
+    const dummyPrev = makeSpan([surrounding[0] || "leftmost"], [], options);
+    const dummyNext = makeSpan([surrounding[1] || "rightmost"], [], options);
+
+    // TODO: These code assumes that a node's math class is the first element
+    // of its `classes` array. A later cleanup should ensure this, for
+    // instance by changing the signature of `makeSpan`.
 
     // Before determining what spaces to insert, perform bin cancellation.
     // Binary operators change to ordinary symbols in some contexts.
-    for (let i = 1; i < nonSpaces.length - 1; i++) {
-        const nonSpacesI: HtmlDomNode = assert(nonSpaces[i]);
-        const left = getOutermostNode(nonSpacesI, "left");
-        if (left.classes[0] === "mbin" &&
-                isBinLeftCanceller(nonSpaces[i - 1], isRealGroup)) {
-            left.classes[0] = "mord";
+    traverseNodes(groups, (node, prev) => {
+        const prevType = prev.classes[0];
+        const type = node.classes[0];
+        if (prevType === "mbin" && utils.contains(binRightCanceller, type)) {
+            prev.classes[0] = "mord";
+        } else if (type === "mbin" && utils.contains(binLeftCanceller, prevType)) {
+            node.classes[0] = "mord";
         }
-
-        const right = getOutermostNode(nonSpacesI, "right");
-        if (right.classes[0] === "mbin" &&
-                isBinRightCanceller(nonSpaces[i + 1], isRealGroup)) {
-            right.classes[0] = "mord";
+    }, {node: dummyPrev}, dummyNext);
+
+    traverseNodes(groups, (node, prev) => {
+        const prevType = getTypeOfDomTree(prev);
+        const type = getTypeOfDomTree(node);
```

Comment:
Why don't we need the `side` argument here?

Reply:
@kevinbarabash The callback function will be only called on non-partial groups.

Reply:
That makes sense, but it's not obvious.  Could you add a comment here before merging this?